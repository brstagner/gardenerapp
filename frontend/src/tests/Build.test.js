import { render, screen, waitFor, fireEvent } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import GardenBuilder from '../gardens/build/GardenBuilder';

const user = { username: "user", userId: "1", isAdmin: "false", token: "user-token" };
const admin = { username: "admin", userId: "2", isAdmin: "true", token: "admin-token" };

// smoke test
it("renders without crashing", function () {
    render(
        <BrowserRouter>
            <GardenBuilder currUser={user} />
        </BrowserRouter>
    );
    // screen.debug;
});

// snapshot test
it("matches snapshot", function () {
    const { asFragment } = render(
        <BrowserRouter>
            <GardenBuilder currUser={user} />
        </BrowserRouter>
    );
    expect(asFragment()).toMatchSnapshot();
});

// buttons
describe("has functional buttons", () => {
    it("adds rows", async function () {
        render(
            <BrowserRouter>
                <GardenBuilder currUser={user} />
            </BrowserRouter>
        );
        const addRow = screen.getByTestId('row-increment');
        fireEvent.click(addRow);
        await waitFor(() =>
            expect(screen.getByText("Rows: 2")).toBeInTheDocument());
    });
    it("deletes rows", async function () {
        render(
            <BrowserRouter>
                <GardenBuilder currUser={user} />
            </BrowserRouter>
        );
        const addRow = screen.getByTestId('row-increment');
        fireEvent.click(addRow);
        const deleteRow = screen.getByTestId('row-decrement');
        fireEvent.click(deleteRow);
        await waitFor(() =>
            expect(screen.getByText("Rows: 1")).toBeInTheDocument());
    });
    it("adds columns", async function () {
        render(
            <BrowserRouter>
                <GardenBuilder currUser={user} />
            </BrowserRouter>
        );
        const addCol = screen.getByTestId('col-increment');
        fireEvent.click(addCol);
        await waitFor(() =>
            expect(screen.getByText("Columns: 2")).toBeInTheDocument());
    });
    it("deletes columns", async function () {
        render(
            <BrowserRouter>
                <GardenBuilder currUser={user} />
            </BrowserRouter>
        );
        const addCol = screen.getByTestId('col-increment');
        fireEvent.click(addCol);
        const deleteCol = screen.getByTestId('col-decrement');
        fireEvent.click(deleteCol);
        await waitFor(() =>
            expect(screen.getByText("Columns: 1")).toBeInTheDocument());
    });
});
