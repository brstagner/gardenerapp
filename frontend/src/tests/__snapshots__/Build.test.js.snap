// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="mb-3 form-floating"
    >
      <input
        autocomplete="name"
        class="form-control"
        id="name"
        name="name"
        placeholder="name"
        type="text"
        value="New Garden"
      />
      <label
        class="form-label"
        for="name"
      >
        Garden Name
      </label>
    </div>
    <div
      class="container-fluid"
    >
      <div
        class="row"
      >
        <div
          class="col"
        >
          <div
            class="container"
          >
            <div
              class="row"
            >
              <div
                class="col"
              >
                <h6>
                  Select a plant from your list by clicking it.
                </h6>
                <h6>
                  Select "REMOVE PLANT" to clear cells.
                </h6>
                <h6>
                  Add that plant to your garden by clicking a cell.
                </h6>
                <div
                  class="row"
                >
                  <div
                    class="col"
                  >
                    <div
                      class="row"
                    >
                      <div
                        class="plant-div col"
                        style="text-align: right; padding: 5px; cursor: pointer;"
                      >
                        REMOVE PLANT
                      </div>
                      <div
                        class="col"
                      />
                    </div>
                  </div>
                  <br />
                  <div
                    class="d-flex justify-content-center"
                  >
                    <button
                      class="btn btn-success"
                    >
                      Get more plants
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="col"
          height="50vh"
        >
          <div
            class="container"
          >
            <h6>
              Add rows and columns:
            </h6>
            <div
              class="row"
            >
              <div
                class="col"
              >
                <div
                  class="text-center"
                >
                  Rows: 1
                </div>
                <div
                  class="buttons d-flex justify-content-center"
                >
                  <button
                    class="btn btn-sm btn-secondary"
                    data-testid="row-decrement"
                  >
                    -
                  </button>
                  <button
                    class="btn btn-sm btn-secondary"
                    data-testid="row-increment"
                  >
                    +
                  </button>
                </div>
              </div>
              <div
                class="col"
              >
                <div
                  class="text-center"
                >
                  Columns: 1
                </div>
                <div
                  class="buttons d-flex justify-content-center"
                >
                  <button
                    class="btn btn-sm btn-secondary"
                    data-testid="col-decrement"
                  >
                    -
                  </button>
                  <button
                    class="btn btn-sm btn-secondary"
                    data-testid="col-increment"
                  >
                    +
                  </button>
                </div>
              </div>
            </div>
          </div>
          <table
            style="width: 100%; text-align: center; cursor: pointer;"
          >
            <tbody>
              <tr>
                <td
                  data-column="1"
                  data-row="1"
                  style="border: 2px solid black; width: 100%; height: 100%; text-overflow: ellipsis;"
                />
              </tr>
            </tbody>
          </table>
          <button
            class="btn btn-primary d-inline-block"
          >
            Save Garden
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
